api/v1alpha1/clusterextension_types.go

```diff
@@ -77,6 +77,7 @@ const (
 	// TODO(user): add more Types, here and into init()
 	TypeInstalled = "Installed"
 	TypeResolved  = "Resolved"
+	TypeDeleting  = "Deleting"
```

Comment:
Could we have a more generic "Progressing" status instead? That way, whether we are in the process of installing, upgrading or deleting, we can set `Progressing=True` (and conversely `Progressing=False` would mean we're at a steady state).

Reply:
Do we need a separate condition for this? I think we could achieve the same thing with `Installed`, with `Installed=True` being our steady state and `False` or `Unknown` for other states, with the `Reason` giving more details.

Reply:
I'm worried if this may cause confusion - what would it mean when: Install is either fail/succeeded, with Progressing - True. How does the user know what action is being taken next? We would have to ensure that when Progressing is set to true, then other status are moved to Unknown.

Reply:
Wondering if we can instead re-name this condition to capture just the `Delete` action similar to how we handle others. 

1. `TypeDelete` Fail - means App's `kcv1alpha1.DeleteFailed` is True, `kcv1alpha1.Deleting` is false.
2. `TypeDelete` Unknown - means App's `kcv1alpha1.Deleting` is true or any other combination which we are unsure of what it represents. 

Looking into how Kapp is handling their status references - https://github.com/carvel-dev/kapp-controller/blob/d89101eb78b2c26a73a7afd8e8ae3111a6731dde/cli/pkg/kctrl/cmd/app/app.go#L39-L64 - All they are trying to map is Reconcile Fail/Succeeded/Progressing, Delete Progressing/Failed. So I'm hoping that the above should capture what we want.

Reply:
> I'm worried if this may cause confusion - what would it mean when: Install is either fail/succeeded, with Progressing - True. How does the user know what action is being taken next? 

We can use the `Reason` and `Message` to get into more specifics.

Taking a stab at enumerating various states:

<table>
<tr>
 <td><b>State</td>
 <td><b>Progressing</td>
 <td><b>Installed</td>
</tr>

<tr>
 <td>Initial install is in progress</td>
 <td>True<br>Reason=Installing<br>Message="installing $bundleID"</td>
 <td>False<br>Reason=NothingInstalled</td>
</tr>

<tr>
 <td>Initial install is failing with a non-terminal error (i.e. retries might resolve the error in the future)</td>
 <td>True<br>Reason=RetryingInstallation<br>Message="retrying installation of $bundleID due to error: $error"</td>
 <td>False<br>Reason=NothingInstalled</td>
</tr>

<tr>
 <td>Initial install failed with a terminal error (i.e. retries will never resolve in the future without some external change)</td>
 <td>False<br>Reason=InstallFailed<br>Message="failed installation of $bundleID: $error</td>
 <td>False<br>Reason=NothingInstalled</td>
</tr>

<tr>
 <td>Install failed, deletion is in progress</td>
 <td>True<br>Reason= Uninstalling <br>Message="uninstalling failed installation of $bundleID"</td>
 <td>False<br>Reason=NothingInstalled</td>
</tr>

<tr>
 <td>Initial install failed, deletion is failing</td>
 <td>True<br>Reason=RetryingUninstall<br>Message="retrying uninstall of $bundleID due to error: $error"</td>
 <td>False<br>Reason=NothingInstalled</td>
</tr>

<tr>
 <td>Initial install failed, deletion failed</td>
 <td>False<br>Reason=UninstallFailed<br>Message="failed uninstall of $bundleID: $error>"</td>
 <td>False<br>Reason=NothingInstalled</td>
</tr>

<tr>
 <td>Install/upgrade succeeded, steady state</td>
 <td>False<br>Reason=NothingToDo</td>
 <td>True<br>Reason=InstallSucceeded<br>Message="successfully installed $bundleID"</td>
</tr>

<tr>
 <td>Install succeeded, upgrade is in progress</td>
 <td>True<br>Reason=Upgrading<br>Message="upgrading to $newBundleID</td>
 <td>True<br>Reason=InstallSucceeded<br>Message="successfully installed $oldBundleID"</td>
</tr>

<tr>
 <td>Install succeeded, upgrade is failing with a non-terminal error</td>
 <td>True<br>Reason=RetryingUpgrade<br>Message="retrying upgrade of $newBundleID due to error: $error"</td>
 <td>True<br>Reason=InstallSucceeded<br>Message="successfully installed $oldBundleID"</td>
</tr>

<tr>
 <td>Install succeeded, upgrade failed with a terminal error</td>
 <td>False<br>Reason=UpgradeFailed<br>Message="failed upgrade to $newBundleID: $error"</td>
 <td>True<br>Reason=InstallSucceeded<br>Message="successfully installed $oldBundleID"</td>
</tr>

<tr>
 <td>Install succeeded, deletion is in progress</td>
 <td>True<br>Reason=Uninstalling<br>Message="uninstalling $bundleID"</td>
 <td>True<br>Reason=InstallSucceeded<br>Message="successfully installed $bundleID"</td>
</tr>

<tr>
 <td>Install succeeded, deletion is failing with a non-terminal error</td>
 <td>True<br>Reason=RetryingUninstall<br>Message="retrying uninstall of $bundleID due to error: $error"</td>
 <td>True<br>Reason=InstallSucceeded<br>Message="successfully installed $bundleID"</td>
</tr>

<tr>
 <td>Install succeeded, deletion is failing with a terminal error</td>
 <td>False<br>Reason=UninstallFailed<br>Message="failed uninstall of $bundleID: $error"</td>
 <td>True<br>Reason=InstallSucceeded<br>Message="successfully installed $bundleID"</td>
</tr>
</table>



Notice how `Progressing` and `Installed` are actually orthogonal. `Installed` does not depend on `Progressing` or vice-versa.

> We would have to ensure that when Progressing is set to true, then other statuses are moved to Unknown.

I don't think so. My idea and goal here is actually that a Progressing condition gives us a place to keep some status about what we're doing so that we don't have to wipe out some status about what we did before. 

For example, if someone installs v1, and then tries to upgrade to v2, but v2 fails the permissions preflight check, the actual state of the system is:
- v1 is still doing its thing untouched
- v2 failed the upgrade preflight check (which might resolve in the future if the admin reconfigures the permissions of the Extension/App's service account).

It would be really nice to be able to say "v1 is still installed and running, but here's why we can't upgrade to v2."

Reply:
Wouldn't a progressing/failing update/delete possibly impact the currently installed extension? If the idea is to have a separate condition specific to preflight checks, we may have to add more support for that separately from the `kapp-controller` App conditions.

Reply:
It might impact it, yes. Hopefully `App`'s status would tell us enough about the failure to help us know if the current installation was impacted or not.

This is probably a good discussion topic with the Carvel folks.